components/gitpod-db/src/typeorm/project-db-impl.ts

```diff
@@ -7,22 +7,35 @@
 import { inject, injectable } from "inversify";
 import { TypeORM } from "./typeorm";
 import { Repository } from "typeorm";
+import { v4 as uuidv4 } from 'uuid';
+import { PartialProject, Project, ProjectEnvVar, ProjectEnvVarWithValue } from "@gitpod/gitpod-protocol";
+import { EncryptionService } from "@gitpod/gitpod-protocol/lib/encryption/encryption-service";
+import { censor } from '@gitpod/gitpod-protocol/lib/util/censor';
 import { ProjectDB } from "../project-db";
 import { DBProject } from "./entity/db-project";
-import { PartialProject, Project } from "@gitpod/gitpod-protocol";
+import { DBProjectEnvVar } from "./entity/db-project-env-vars";
+
+function censorEnvVar(variable: ProjectEnvVarWithValue): ProjectEnvVar {
```

Comment:
IMO this is a clear "service" concern, and does not belong into the DB layer.
But maybe I'm missing the motivation here... And: why was this not necessary for Users? :thinking:

Reply:
Agree, this is a new practice: Instead of passing around potentially secret values through the entire meta code base (where it has many opportunities to get logged accidentally), I prefer to pass around a "handle" (which can get logged without consequence) and only use the actual value where strictly necessary (you'll notice that the dashboard and the API can never read values, only set or delete them).

This is still far from perfect, and in the future we'll likely want to use End-To-End encryption and Kubernetes secrets (thus passing a no-value "handle" further down to workspaces) but with this approach, you now have just a single place in meta that reads values (WorkspaceStarter.createSpec) as opposed to many places.

Reply:
Ah, now I see what you want to achieve. I mis-read the types, and got confused by `censor`: Basically the `SELECTED` fields should match the type (precisely).

I recommend changing the function name then, maybe prefix with `filter`, or so.

Alternatively, you can "fix" the `select` in the first place. Here's an example: https://github.com/gitpod-io/gitpod/blob/7abe78c91792c18b0c22103c890b25ea3a4cdf05/components/gitpod-db/src/typeorm/workspace-cluster-db-impl.ts#L57

Reply:
Thanks! I really like the `.select` suggestion, but feel it would make many `repo.find` look much more complicated than they really are.

If `censorEnvVar` doesn't work, I'll rename the function to `toProjectEnvVar` (similar to other `toShape` functions in other DB impls).

I personally would find the word `filter` confusing here, as to me `filter` is strongly associated with JS list filtering (i.e. I would automatically assume a `filterProjectEnvVar` function takes a list of `ProjectEnvVar`s and returns a sub-list based on some criteria -- which is not the case here).

Reply:
`toProjectEnvVar` is fine with me as well. :+1: